### PR TITLE
Add ContributeAndFinalize task detector

### PR DESCRIPTION
**Pull request by mcornaton**

### CODE REVIEW

File: src/main/java/com/iexec/core/detector/task/FinalizedTaskDetector.java [hunk 1]
```diff
@@ -37,30 +41,68 @@ public class FinalizedTaskDetector implements Detector {
...
+        log.debug("Trying to detect contributed and finalized tasks");
+        taskService.findByCurrentStatus(TaskStatus.RUNNING)
+                .stream()
+                .filter(this::isTaskContributeAndFinalizeDone)
+                .forEach(this::publishTaskUpdateRequest);
+    }
+
+    boolean isChainTaskCompleted(Task task) {
+        final Optional<ChainTask> chainTask = iexecHubService.getChainTask(task.getChainTaskId());
+        return chainTask.isPresent() && chainTask.get().getStatus().equals(ChainTaskStatus.COMPLETED);
```
Comment (jbern0rd) [inline src/main/java/com/iexec/core/detector/task/FinalizedTaskDetector.java:88]:
```suggestion
        return chainTask.isPresent() && chainTask.get().getStatus() == ChainTaskStatus.COMPLETED;
```

File: src/test/java/com/iexec/core/detector/task/FinalizedTaskDetectorTests.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+        when(taskService.findByCurrentStatus(TaskStatus.FINALIZING)).thenReturn(List.of(task));
+
+        detector.detectFinalizedTasks();
+
+        verify(taskUpdateRequestManager).publishRequest(CHAIN_TASK_ID);
+    }
+
+    @Test
+    void shouldDetectNoFinalizedTassAsTaskIsRevealing() {
```
[the remark this answers is not in the document]
Comment (mcornaton) [reply src/test/java/com/iexec/core/detector/task/FinalizedTaskDetectorTests.java]: Please see https://github.com/iExecBlockchainComputing/iexec-core/pull/593/commits/43f8782e8acade6649b8692176092feac2cf64f3.

File: src/test/java/com/iexec/core/detector/task/FinalizedTaskDetectorTests.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
+package com.iexec.core.detector.task;
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/test/java/com/iexec/core/detector/task/FinalizedTaskDetectorTests.java:17]: Thanks !

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,7 +10,7 @@ All notable changes to this project will be documented in this file.
 - Add `running2Finalized2Completed` in `TaskUpdateManager`. (#577 #578)
 - Disable `contributeAndFinalize` with CallBack. (#579 #581)
 - Add purge cached task descriptions ability. (#587)
-- Add `ContributionAndFinalizationUnnotifiedDetector`. (#590)
+- Add detectors for ContributionAndFinalizeFlow. (#590 #593)
```
Comment (jbern0rd) [inline CHANGELOG.md:13]:
```suggestion
- Add detectors for ContributeAndFinalizeFlow. (#590 #593)
```
or
```suggestion
- Add detectors for `ContributeAndFinalize` flow. (#590 #593)
```
What do you think ?

Comment (mcornaton) [reply CHANGELOG.md]: I prefer the second version. Thanks! :)

**Review by jbern0rd — APPROVED** [review]

**Pull request merged**